### DISCUSSION TITLE
SDK SSO Login Refresh Token

### DISCUSSION THREAD

**Discussion by NilsBaumgartner1994** [directus/directus#22427] [Feature Requests]
### Summary

So as of https://github.com/directus/directus/pull/21239 the SSO login method changed to session by default: 
```
Updated SSO (open-id, saml, oath2) to default to the new "session" mode
```

How to revert this? Or how can i login in my native app via SSO using the sdk?
My current approach was to redirect to the app and passing a refresh_token, which currently no longer is possible since i only get the "directus_session_token"?

Maybe also consider updating the Doc: https://docs.directus.io/reference/authentication.html#request
To add a hint, that SSO has a different mode.

### Motivation

The support for SSO login in native apps is rising as of multiple issues listed.

### Detailed Design

As there is a security issue exposing the token in the URL i would propose to add a whitelist of allowred redirects with exposing the refresh_token in the url, so that native apps using deep link can access it.
An in-App Browser is also not a secure option, as it could be mocked.

### Must Have:
- Improved Documentation hinting that SSO logins are using a different mode.

### Could Have:
- Elaborating a solution for SDK-JS for SSO logins supporting native apps

### Drawbacks

-

### Alternatives

-

### Adoption Strategy

-

**Comment by NilsBaumgartner1994** [comment]
Current hacky solution to pass `refresh_token` to native app users, as `session`and `cookie` do not work.
Please keep in mind:
- Check redirect url if it is allowed
- Clear the refresh_token when not read:

Create custom endpoint:

```
import {defineEndpoint} from '@directus/extensions-sdk';
import ms from 'ms';

// TO Test this Endpoint:
// 1. Login with a user in the Directus Admin UI
// 2. Go to the URL: http://127.0.0.1/rocket-meals/api/redirect-with-token?redirect=http://localhost:8081/login?directus_refresh_token=
// Where http://127.0.0.1/rocket-meals/api is the URL of the Directus API
export default defineEndpoint({
	id: 'redirect-with-token', // this will be the URL at which this endpoint is accessible
	handler: (router, {
		services,
		database,
		env,
		getSchema,
		logger
	}) => {
		router.get('/', async (req, res) => {
			let redirectUrlIsValid = true;
			const redirect = req.query.redirect;

			//TODO: check if redirect is in listOfAllowedRedirects

			if(!redirectUrlIsValid){
				res.status(400).send("Invalid redirect URL");
				return;
			}

			const directus_refresh_token = req.cookies.directus_refresh_token;

			if(!!directus_refresh_token){ // this means the auth provider is using "cookie" or "json" mode.
				const redirectURL = redirect + directus_refresh_token;
				res.redirect(redirectURL);
				return;
			}

			const directus_session_token = req.cookies.directus_session_token;
			if(!!directus_session_token) { // this means the auth provider is using "session" mode.
				// we need to obtain the directus_refresh_token from the directus_session_token
				const accountability = req?.accountability;
				const userId = req?.accountability?.user;
				if(!userId){
					res.status(400).send("No user found");
					return;
				}
				
				const knex = database;
				const { nanoid } = await import('nanoid');

				const refreshToken = nanoid(64);
				const msRefreshTokenTTL: number = ms(String(env['REFRESH_TOKEN_TTL'])) || 0;
				const refreshTokenExpiration = new Date(Date.now() + msRefreshTokenTTL);

				await knex('directus_sessions').insert({
					token: refreshToken,
					user: userId,
					expires: refreshTokenExpiration,
					ip: accountability?.ip,
					user_agent: accountability?.userAgent,
					origin: accountability?.origin,
				});

				await knex('directus_sessions').delete().where('expires', '<', new Date());

				const redirectURL = redirect + refreshToken;
				res.redirect(redirectURL);
				return;
			}

			// no token found
			res.status(400).send("No token found");
		});
	}
});
´´´

**Reply by NilsBaumgartner1994** [reply]
As this implicit flow is not recommended, as the token could be grabed by eavedropping here is a possible PKCE implementation:
Have a look at our PKCE solution?: https://github.com/directus/directus/discussions/23611

**Reply by juhroque** [reply]
Hey! Any updates on this? Have you found another way to use the refresh token with the session cookie?

**Reply by NilsBaumgartner1994** [reply]
> Hey! Any updates on this? Have you found another way to use the refresh token with the session cookie?

Hey, thanks for the interest. We are still using the Feature abouve. There were discussions about this feature but to best of my knowledge it's not officially supported otherwise.
